h2o-r/h2o-package/R/frame.R

```diff
@@ -3395,49 +3431,62 @@ as.h2o.Matrix <- function(x, destination_frame="", ...) {
   .fetch.data(x,1L)
-  # Versions of R prior to 3.1 should not use hex string.
-  # Versions of R including 3.1 and later should use hex string.
-  use_hex_string <- getRversion() >= "3.1"
-
-  urlSuffix <- paste0('DownloadDataset',
-                      '?frame_id=', URLencode( h2o.getId(x)),
-                      '&hex_string=', as.numeric(use_hex_string))
-  
-  verbose <- getOption("h2o.verbose", FALSE)
-  if (verbose) pt <- proc.time()[[3]]
-  ttt <- .h2o.doSafeGET(urlSuffix = urlSuffix)
-  if (verbose) cat(sprintf("fetching from h2o frame to R using '.h2o.doSafeGET' took %.2fs\n", proc.time()[[3]]-pt))
-  n <- nchar(ttt)
-
-  # Delete last 1 or 2 characters if it's a newline.
-  # Handle \r\n (for windows) or just \n (for not windows).
-  chars_to_trim <- 0L
-  if (n >= 2L) {
-    c <- substr(ttt, n, n)
-    if (c == "\n") chars_to_trim <- chars_to_trim + 1L
-    if (chars_to_trim > 0L) {
-      c <- substr(ttt, n-1L, n-1L)
-      if (c == "\r") chars_to_trim <- chars_to_trim + 1L
-    }
-  }
-
-  if (chars_to_trim > 0L) {
-    ttt2 <- substr(ttt, 1L, n-chars_to_trim)
-    ttt <- ttt2
-  }
-
+    
   # Get column types from H2O to set the dataframe types correctly
   colClasses <- attr(x, "types")
   colClasses <- gsub("numeric", NA, colClasses) # let R guess the appropriate numeric type
-  colClasses <- gsub("int", NA, colClasses) # let R guess the appropriate numeric type
+  colClasses <- gsub("int", NA, colClasses) # let R guess the appropriate numeric typepodobne jako mame
   colClasses <- gsub("real", NA, colClasses) # let R guess the appropriate numeric type
   colClasses <- gsub("enum", "factor", colClasses)
   colClasses <- gsub("uuid", "character", colClasses)
   colClasses <- gsub("string", "character", colClasses)
   colClasses <- gsub("time", NA, colClasses) # change to Date after ingestion
-  
+
   # Convert all date columns to POSIXct
   dates <- attr(x, "types") %in% "time"
+    
+  nCol <- attr(x, "ncol")
+  nRow <- attr(x, "nrow")
+    
+  # Due to data.frame limitation of vector size, only smaller data dimension than .Machine$integer.max are allowed
+  if(nCol > .Machine$integer.max || nRow > .Machine$integer.max){
+    stop("It is not possible convert H2OFrame to data.frame/data.table. The H2OFrame is bigger than vector size limit for R.")  
+  }  
+    
+  # Versions of R prior to 3.1 should not use hex string.
+  # Versions of R including 3.1 and later should use hex string.
+  useHexString <- getRversion() >= "3.1"
+
+  urlSuffix <- paste0('DownloadDataset',
+                      '?frame_id=', URLencode( h2o.getId(x)),
+                      '&hex_string=', as.numeric(useHexString))
+  
+  verbose <- getOption("h2o.verbose", FALSE)
+    
+  if (verbose) pt <- proc.time()[[3]]
+  
+  # Get data in binary format for case the data are too big to load in character format  
+  payload <- .h2o.doSafeGET(urlSuffix = urlSuffix, binary=TRUE)
+  
+  if(length(payload) < .Machine$integer.max)  {
```

Comment:
Good idea! I think it would be good to parametrize this in order to be able to use this code path even for smaller datasets (for testing purposes).

    getOption("h2o.as.data.frame.max.in-memory.payload.size", .Machine$integer.max) 

something like this?